---
title: "ChIPseq_230727"
author: "Andrea"
date: "2023/08/02"
output: html_document
---

fastq location: /pipeline/Runs/NextSeq/230727_VH01624_8_AAC3HG5HV/ProjectFolders/Project_Jesse-Balic/Sample_JB[1][8-9]|Sample_JB[2][0-4]
project directory: /dawson_genomics/Projects/zebrafish/ChIPseq/230727_VH01624_8_AAC3HG5HV/

# set up project dir and process (fastqc, trim, align, tdfs)
```{bash}
mkdir alignment
mkdir fastqc
mkdir trim_fastq
mkdir logs
mkdir scripts
mkdir session-info

# many samples in this run, limit this to those for Vicky
for fq in /pipeline/Runs/NextSeq/230727_VH01624_8_AAC3HG5HV/ProjectFolders/Project_Jesse-Balic/Sample_JB[1][8-9]/*_R1_001.fastq.gz
do
sbatch scripts/alignHg38.sbatch $fq
done

for fq in /pipeline/Runs/NextSeq/230727_VH01624_8_AAC3HG5HV/ProjectFolders/Project_Jesse-Balic/Sample_JB[2][0-4]/*_R1_001.fastq.gz
do
sbatch scripts/alignHg38.sbatch $fq
done

module load multiqc/1.8
multiqc .
```

# peak calling (both narrow and broad) include bedgraph output for Spark plots and if diff peaks is needed
```{bash}
mkdir macs/narrowpeaks
mkdir macs/broadpeaks

sbatch scripts/macs2.sbatch alignment/JB18_S36.sort.rmdup.bam alignment/JB24_S43.sort.rmdup.bam
sbatch scripts/macs2.sbatch alignment/JB19_S37.sort.rmdup.bam alignment/JB24_S43.sort.rmdup.bam
sbatch scripts/macs2.sbatch alignment/JB20_S39.sort.rmdup.bam alignment/JB24_S43.sort.rmdup.bam
sbatch scripts/macs2.sbatch alignment/JB21_S40.sort.rmdup.bam alignment/JB24_S43.sort.rmdup.bam
sbatch scripts/macs2.sbatch alignment/JB22_S41.sort.rmdup.bam alignment/JB24_S43.sort.rmdup.bam
sbatch scripts/macs2.sbatch alignment/JB23_S42.sort.rmdup.bam alignment/JB24_S43.sort.rmdup.bam
```

# combine reps, then use bdgdiff to compare peaks 
### no need for depth adj as all samples in each experiment used same input as shown above, individual control bedgraphs made for consistency
```{bash}
# quick and low mem usage so can be run interactively
module load macs/2.2.7.1

cd macs/broadpeaks
mkdir macs_diff

macs2 cmbreps -i JB18_S36_treat_pileup.bdg JB19_S37_treat_pileup.bdg JB20_S39_treat_pileup.bdg -m mean -o HepG2_untreated_treat_pileup.bdg
macs2 cmbreps -i JB21_S40_treat_pileup.bdg JB22_S41_treat_pileup.bdg JB23_S42_treat_pileup.bdg -m mean -o HepG2_APAP_24h_treat_pileup.bdg
macs2 cmbreps -i JB18_S36_control_lambda.bdg JB19_S37_control_lambda.bdg JB20_S39_control_lambda.bdg -m mean -o HepG2_untreated_control_lambda.bdg
macs2 cmbreps -i JB21_S40_control_lambda.bdg JB22_S41_control_lambda.bdg JB23_S42_control_lambda.bdg -m mean -o HepG2_APAP_24h_control_lambda.bdg

macs2 bdgdiff --t1 HepG2_APAP_24h_treat_pileup.bdg --c1 HepG2_APAP_24h_control_lambda.bdg --t2 HepG2_untreated_treat_pileup.bdg --c2 HepG2_untreated_control_lambda.bdg --o-prefix HepG2_APAP_24hvUntreated --outdir macs_diff


cd ../narrowpeaks
mkdir macs_diff

macs2 cmbreps -i JB18_S36_treat_pileup.bdg JB19_S37_treat_pileup.bdg JB20_S39_treat_pileup.bdg -m mean -o HepG2_untreated_treat_pileup.bdg
macs2 cmbreps -i JB18_S36_treat_pileup.bdg JB19_S37_treat_pileup.bdg JB20_S39_treat_pileup.bdg -m mean -o HepG2_APAP_24h_treat_pileup.bdg
macs2 cmbreps -i JB18_S36_control_lambda.bdg JB19_S37_control_lambda.bdg JB20_S39_control_lambda.bdg -m mean -o HepG2_untreated_control_lambda.bdg
macs2 cmbreps -i JB21_S40_control_lambda.bdg JB22_S41_control_lambda.bdg JB23_S42_control_lambda.bdg -m mean -o HepG2_APAP_24h_control_lambda.bdg

macs2 bdgdiff --t1 HepG2_APAP_24h_treat_pileup.bdg --c1 HepG2_APAP_24h_control_lambda.bdg --t2 HepG2_untreated_treat_pileup.bdg --c2 HepG2_untreated_control_lambda.bdg --o-prefix HepG2_APAP_24hvUntreated --outdir macs_diff
```
