---
title: "Vicky SLAM-Seq Analysis 2021 October"
output: html_notebook
---

# Experimental design:
Zebrafish WT_UPRT and Nrf2KO_UPRT with and without APAP treatment
Control WT CFP only with DMSO
All samples spiked in with S2 cells
Information files under sample_info folder

# Location of files:
Run:
/pipeline/Runs/NextSeq/210922_NB501056_0694_AHJJHWBGXK/ProjectFolders/Project_Vicky-Tan/

# Setting up references and running SLAM-Seq pipeline

```{r}
dir.create("reference_files")
```

## Dm normalisation with slam-seq

Get (S2 cell spike in) Dm (Drosophila_melanogaster - fruit fly) 3'UTR coordinates
```{r, eval = FALSE}
library(biomaRt)
Dm6ensembl=useMart(host='asia.ensembl.org', biomart='ENSEMBL_MART_ENSEMBL', dataset='dmelanogaster_gene_ensembl')
Dm6enz = read.delim("reference_files/Drosophila_melanogaster_BDGP6.32.txt",header=T) #file generated from ensembl website
Dm6FBgn = Dm6enz[!duplicated(Dm6enz[,1]),] #removing duplicates

utr = getBM(attributes = c('chromosome_name','3_utr_start','3_utr_end','ensembl_gene_id','external_gene_name','strand','ensembl_transcript_id'), filters = 'ensembl_gene_id',values = Dm6FBgn$Gene.stable.ID, mart = Dm6ensembl)

utrna = utr[!is.na(utr$`3_utr_start`),]
utrna$strand2=ifelse(utrna$strand=="-1","-","+")

write.table(utrna[,c(1:5,8)],"reference_files/Dm6allutr.bed",sep="\t",quote=F,row.names = F,col.names = F)
```

```{bash, eval = F}
cd reference_files
sort -k1,1 -k2,2n Dm6allutr.bed > Dm6allutrsort.bed
module load bedtools # for running on cluster
bedtools merge -c 4,5,6 -o distinct,distinct,distinct -s -i Dm6allutrsort.bed > Dm6allutrmerge.bed
```

Changing chromosome prefix
```{r, eval = FALSE}
utr = read.delim("reference_files/Dm6allutrmerge.bed",header=F)
table(utr$V1)
utr$V1=as.character(utr$V1)
utr$V1 <- paste("DM", utr$V1, sep = "_" ) 
table(utr$V1)

utr$Ens = unlist(lapply(strsplit(as.character(utr$V5), ","), '[[', 1))
utr$Gene = unlist(lapply(strsplit(as.character(utr$V6), ","), '[[', 1))

dm6allutrmergechr = utr[,c(1:3,8,9,7)]
names(dm6allutrmergechr) = c("chr", "start", "end", "ensembl", "gene.id", "strand")

write.table(dm6allutrmergechr,"reference_files/Dm6allutrmergechr.bed",sep="\t",quote=F,row.names = F,col.names = F)
```


## Dr (Danio rerio - Zebrafish) UTR for slam-seq

Get Dr 3'UTR coordinates
```{r, eval = FALSE}
library(biomaRt)
Drensembl <- useMart(
  host = "https://asia.ensembl.org", 
  biomart = 'ENSEMBL_MART_ENSEMBL',
  dataset = 'drerio_gene_ensembl'
  )
GRCz11ens = read.delim("reference_files/GRCz11_genes.txt",header=T)
GRCz11ens = GRCz11ens[!duplicated(GRCz11ens[,1]),]

Dr_utr = getBM(
  attributes = c('chromosome_name','3_utr_start','3_utr_end','ensembl_gene_id','external_gene_name','strand','ensembl_transcript_id'), 
  filters = 'ensembl_gene_id',
  values = GRCz11ens$Gene.stable.ID, 
  mart = Drensembl
  )

utrna = Dr_utr[!is.na(Dr_utr$`3_utr_start`),]
utrna$strand2=ifelse(utrna$strand=="-1","-","+")

write.table(
  utrna[,c(1:5,8)],
  "reference_files/GRCz11_allutr.bed",
  sep="\t",
  quote=F,
  row.names = F,
  col.names = F
  )
```

In terminal
```{bash, eval = F}
cd reference_files
sort -k1,1 -k2,2n GRCz11_allutr.bed > GRCz11_allutrsort.bed
module load bedtools
# bedtools merge -c 4,5,6 -o distinct,distinct,distinct -s -i GRCz11_allutrsort.bed > GRCz11_allutrmerge.bed
```

```{r, eval = FALSE}
utr = read.delim("reference_files/GRCz11_allutrsort.bed",header=F)
utr
ens_unique = unique(utr$V5)

utr[utr$V5 == ens_unique[2],]
min(utr[utr$V5 == ens_unique[2],]$V2)
max(utr[utr$V5 == ens_unique[2],]$V3)

utr[utr$V5 == ens_unique[7],]
min(utr[utr$V5 == ens_unique[7],]$V2)
max(utr[utr$V5 == ens_unique[7],]$V3)

collapseUTR = function(utr = utr, ens_unique = ens_unique){
output = data.frame()
for (l in 1:length(ens_unique)){
  genel = data.frame("chr" = min(utr[utr$V5 == ens_unique[l],]$V1),
    "start" = min(utr[utr$V5 == ens_unique[l],]$V2),
    "end" = max(utr[utr$V5 == ens_unique[l],]$V3),
    "ensembl" = min(utr[utr$V5 == ens_unique[l],]$V4),
    "gene id"= min(utr[utr$V5 == ens_unique[l],]$V5),
    "strand" = min(utr[utr$V5 == ens_unique[l],]$V6))
  output = rbind(output, genel)}
return(output)}

Dr_utr_collapse = collapseUTR(utr, ens_unique)

Dr_utr_collapse
write.table(Dr_utr_collapse,"reference_files/GRCz11_allutrmerge.bed",sep="\t",quote=F,row.names = F,col.names = F)

# utr$Ens = unlist(lapply(strsplit(as.character(utr$V5), ","), '[[', 1))
# utr$Gene = unlist(lapply(strsplit(as.character(utr$V6), ","), '[[', 1))

# write.table(utr[,c(1:3,8,9,7)],"GRCz11_allutrmergechr.bed",sep="\t",quote=F,row.names = F,col.names = F)
```

Combine Dm6 and GRCz11 utr bed files
```{r, eval = FALSE}
GRCz11utr = read.delim("reference_files/GRCz11_allutrmerge.bed",header=F)
Dm6utr = read.delim("reference_files/Dm6allutrmergechr.bed", header=F)
GRCz11Dm6utr = rbind(GRCz11utr,Dm6utr)
write.table(GRCz11Dm6utr,"/data/reference/dawson_labs/joint_genomes/Dm6Dr11/GRCz11Dm6allutrmerge.bed",sep="\t",quote=F,row.names = F,col.names = F)
```

# copy to and rehead dm6 fasta in home dir (to protect integrity of reference), then create joint genome
```{bash, eval = F}
cp /data/reference/indexes/fruitfly/dm6/fasta/Drosophila_melanogaster.BDGP6.dna.toplevel.fa /home/agillespie/Drosophila_melanogaster.BDGP6.dna.toplevel.fa

sed 's/^>/>DM_/g' /home/agillespie/Drosophila_melanogaster.BDGP6.dna.toplevel.fa > /home/agillespie/Drosophila_melanogaster.BDGP6.dna.toplevel.reheaded.fa

cat /home/agillespie/Drosophila_melanogaster.BDGP6.dna.toplevel.reheaded.fa /data/reference/indexes/zebrafish/danrer11/fasta/Danio_rerio.GRCz11.dna.primary_assembly.fa > /data/reference/dawson_labs/joint_genomes/Dm6Dr11/Dm6Dr11.fa
```

Optional to add mcherry sequence (might be useful to have)

# add mCherry fasta to joint genome and bed to combined utr
```{bash, eval = F}
cat /data/reference/dawson_labs/joint_genomes/Dm6Dr11/Dm6Dr11.fa /dawson_genomics/Projects/zebrafish/SLAMSeq/2106_SLAMSeq/ref_wCherry/mCherry.fa > /data/reference/dawson_labs/joint_genomes/Dm6Dr11/Dm6Dr11mCherry.fa

cat GRCz11Dm6allutrmerge.bed /dawson_genomics/Projects/zebrafish/SLAMSeq/2106_SLAMSeq/ref_wCherry/mCherry_utr.bed > GRCz11Dm6mCherryallutrmerge.bed
```


slam.sbatch file
```{bash, eval = FALSE}
#!/bin/bash
 
#SBATCH --nodes=1
#SBATCH --ntasks=1
#SBATCH --mem-per-cpu=40G
#SBATCH --time=0-12:00:00
#SBATCH --output=../logs/slam%j.stdout
#SBATCH --error=../logs/slam%j.err
#SBATCH --mail-user=andrea.gillespie@petermac.org
#SBATCH --mail-type=END
#SBATCH --job-name="slam"

module load slamdunk
module load fastqc

bname=`basename $1 _R1_001.fastq.gz`
rl=75
Ref=/data/reference/dawson_labs/joint_genomes/Dm6Dr11/Dm6Dr11mCherry.fa
utr=/dawson_genomics/Projects/zebrafish/SLAMSeq/210922_NB501056_0694_AHJJHWBGXK/GRCz11Dm6mCherryallutrmerge.bed

mkdir ${bname}

java -jar /config/binaries/trimmomatic/0.39/trimmomatic-0.39.jar SE -threads 1 -phred33 $1 ${bname}/${bname}.trim.fq.gz ILLUMINACLIP:/config/binaries/trimmomatic/0.39/adapters/TruSeq3-SE.fa:2:30:10 LEADING:3 TRAILING:3 SLIDINGWINDOW:4:15 MINLEN:36

#fastqc $1

fastqc ${bname}/${bname}.trim.fq.gz

srun -n 1 slamdunk all -r $Ref -b $utr -fb $utr -o ${bname}slamout -5 12 -n 100 -t 8 -m -mv 0.2 -mts -mbq 27 -rl $rl ${bname}/${bname}.trim.fq.gz

srun -n 1 alleyoop collapse -o collapse/ -t 1 ${bname}slamout/count/*tcount.tsv &

srun -n 1 alleyoop snpeval -o ${bname}slamout/ -s ${bname}slamout/snp/ -r $Ref -b $utr -l $rl -t 1 ${bname}slamout/filter/*_slamdunk_mapped_filtered.bam &
srun -n 1 alleyoop rates -o ${bname}slamout/ -r $Ref -t 1 ${bname}slamout/filter/*_slamdunk_mapped_filtered.bam &
srun -n 1 alleyoop tccontext -o ${bname}slamout/ -r $Ref -t 1 ${bname}slamout/filter/*_slamdunk_mapped_filtered.bam &
srun -n 1 alleyoop tcperutrpos -o ${bname}slamout/ -r $Ref -b $utr -s ${bname}slamout/snp/ -l $rl -t 1 ${bname}slamout/filter/*_slamdunk_mapped_filtered.bam &
srun -n 1 alleyoop tcperreadpos -o ${bname}slamout/ -r $Ref -s ${bname}slamout/snp/ -l $rl -t 1 ${bname}slamout/filter/*_slamdunk_mapped_filtered.bam &
srun -n 1 alleyoop utrrates -o ${bname}slamout/ -r $Ref -l $rl -t 1 -b $utr ${bname}slamout/filter/*_slamdunk_mapped_filtered.bam &

wait
```

Run 1
Slam-seq samples through slam.sbatch from alignment dir
```{bash, eval = F}
cd alignment

for fq in /pipeline/Runs/NextSeq/210922_NB501056_0694_AHJJHWBGXK/ProjectFolders/Project_Vicky-Tan/*/*.fastq.gz
do sbatch ../slamGRCz11210922.sbatch $fq
done
```

# run multiqc on fastqc and slamdunk files
```{bash}
# still in aligment dir to run multiqc
module load multiqc/1.8
multiqc .
```

# After the alignments are done there is a collapse folder with all the counts

# Change collapse .csv files to .tsv
```{bash, eval = F}
cd collapse
mkdir tsv
for file in *.csv; do cp "$file" tsv/"${file%.csv}.tsv"; done
```

# Read in slam-dunk collapse counts files

Slamdunk Tcount file format:
col1 is entrez geneID
col3 is RPM
col8 is total count
col9 is TC count

Loading and reading in files
```{r}
# set dirs for analysis and create as needed
project.dir <- "/dawson_genomics/Projects/zebrafish/SLAMSeq/210922_NB501056_0694_AHJJHWBGXK/"
alignment.dir <- file.path(project.dir, 'alignment')
tsv.dir <- file.path(alignment.dir, 'collapse/tsv')
plots.dir <- file.path(project.dir, 'plots')
results.dir <- file.path(project.dir, 'results')

dir.create(results.dir)
dir.create(plots.dir)

setwd(project.dir)

library(edgeR)
files <- list.files(tsv.dir, pattern="*.tsv$")
sample.sheet <- X2021_09_03_Whole_Larvae_SLAMitseq
# files
samples_mch = data.frame(
  "files" = files, 
  "samples" = gsub(".trim.fq_slamdunk_mapped_filtered_tcount_collapsed.tsv", "", files), 
  "time" = gsub("h", "", sample.sheet$Other_Conditions), 
  "treatment" = sample.sheet$Treatment,
  "cells" = sample.sheet$`Cell type/line`
  )

# sequencing and sample information
counts_mch = readDGE(
  file = files, 
  path = tsv.dir, 
  columns=c(1,8), 
  group=samples_mch$samples
  )

TCcounts_mch = readDGE(
  file = files,
  path = tsv.dir,
  columns=c(1,9), 
  group=samples_mch$samples
  )

RPM_mch = readDGE(
  files = files,
  path = tsv.dir,
  columns=c(1,3), 
  group=samples_mch$samples
  )

# cleaning up names
row.names(counts_mch$samples) <- gsub(".trim.fq_slamdunk_mapped_filtered_tcount_collapsed", "", row.names(counts_mch$samples))
row.names(TCcounts_mch$samples) <- gsub(".trim.fq_slamdunk_mapped_filtered_tcount_collapsed", "", row.names(TCcounts_mch$samples))
colnames(counts_mch$counts) <- gsub(".trim.fq_slamdunk_mapped_filtered_tcount_collapsed", "", colnames(counts_mch$counts))
colnames(TCcounts_mch$counts) <- gsub(".trim.fq_slamdunk_mapped_filtered_tcount_collapsed", "", colnames(TCcounts_mch$counts))
```

# write out raw counts
```{r}
raw_counts = counts_mch$counts
raw_TCcounts = TCcounts_mch$counts

write.table(
  raw_counts, 
  "results/counts_raw_all.txt",
  sep="\t",
  quote=F
  )

write.table(
  raw_TCcounts, 
  "results/TCcounts_raw_all.txt",
  sep="\t",
  quote=F
  )
```

# Probably a good place to separate out the zebrafish and drosphila reads
# Then analyse the drosphila and zebrafish separatly until the normalisation.
# apply the normalisation factors from the drosophila to the zebrafish, then continue with DE with or without the drosophila normalisation factor as a comparison.

# separate organisms counts objects and make list for looping
```{r}
dros <- which(substr(rownames(counts_mch), 1, 4) == "FBgn")

dros.counts <- counts_mch[dros, ]
zeb.counts <- counts_mch[-dros, ]
dros.TCcounts <- TCcounts_mch[dros, ]
zeb.TCcounts <- TCcounts_mch[-dros, ]

counts.obs <- list(zeb.counts, dros.counts, zeb.TCcounts, dros.TCcounts)
```

# Filtering out counts of genes with low counts
```{r}
# init variable for cpms list
counts.cpms.list <- list()

# loop through both counts objects to apply same filtering to both
for(counts.i in 1:length(counts.obs)){
  colSums(counts.obs[[counts.i]]$counts)
  noint = rownames(counts.obs[[counts.i]]) %in% c("__no_feature", "__ambiguous", "__too_low_aQual", "__not_aligned", "__alignment_not_unique")
  table(noint)
  counts.cpms.list[[counts.i]] = cpm(counts.obs[[counts.i]])
  keep = rowSums(counts.cpms.list[[counts.i]] >1) >=3 & !noint
  counts.obs[[counts.i]] = counts.obs[[counts.i]][keep,]
  dim(counts.obs[[counts.i]]$counts)
}

# set names for lists
names(counts.obs) <- c("zeb_counts", "dros_counts", "zeb_TCcounts", "dros_TCcounts")
names(counts.cpms.list) <- names(counts.obs)
```


Save and annotate zebrafish count files
```{r}
counts.names <- c(rep("counts", 2), rep("TCcounts", 2))

for(i in c(1, 3)){
  write.table(
    counts.obs[[i]]$counts,
    file.path(results.dir, paste0(counts.names[[i]], "_filtered_zebrafish_mCherry.txt")),
    sep="\t",
    quote=F
    )

  # Zebrafish annotation
  zfish_genes=getBM(
    attributes=c("ensembl_gene_id","zfin_id_symbol","entrezgene_id"),
    filters = "ensembl_gene_id",
    values=rownames(counts.cpms.list[[i]]),
    mart=Drensembl
    )
  zfish_genesnodup = zfish_genes[!duplicated(zfish_genes$ensembl_gene_id),]
  EnsGenes = data.frame(
    "zeb.cpm.ensg" = rownames(counts.cpms.list[[i]])
    )
  genesAnno <- left_join(
    EnsGenes,
    zfish_genesnodup,
    by=c("zeb.cpm.ensg"="ensembl_gene_id")
    )
  colnames(genesAnno)=c("zfish_ensembl_gene_id","zfin_id_symbol","zfish_entrez")
  counts.cpms.list[[i]] = as.data.frame(counts.cpms.list[[i]])
  counts.cpms.list[[i]]$zfin_id_symbol <- genesAnno$zfin_id_symbol

  write.table(
    counts.cpms.list[[i]],
    file.path(results.dir, paste0(counts.names[[i]], "_cpms_filtered_zebrafish_mCherry.txt")),
    sep = '\t',
    quote = F
    )  
}
```

Read saved counts
```{r}
counts = read.delim(file = "results/counts_filtered_zebrafish_mCherry.txt",sep="\t")
TCcounts = read.delim(file = "results/TCcounts_filtered_zebrafish_mCherry.txt",sep="\t")
```

# Plot count numbers
```{r}
# set names for consistency across samples and readability
names.arg <- paste(sample.sheet$`Submitted Sample ID`, paste0("S", 1:27), sep = "_")
pdf(file = file.path(plots.dir, "barplots_summary.pdf"), width = 6, height = 4)
barplot(colSums(counts), las=2, main="total reads", names.arg = names.arg)
barplot(colSums(TCcounts), las=2, main="TC reads", names.arg = names.arg)
barplot(colSums(TCcounts)/colSums(counts)*100, las=2, main="percentage TC", names.arg = names.arg)
dev.off()
```

# make table of TC%
```{r}
TC.table <- data.frame(
  "sample.names" = names.arg, 
  "percent.TC" = colSums(TCcounts)/colSums(counts)*100
  )
write.table(
  TC.table, 
  file.path(results.dir, "all_sample_percent_TC.txt"),
  sep ="\t",
  col.names = T,
  row.names = F,
  quote = F
  )
```

# Making DGE list object and setting design table

## counts
```{r}
group = paste(samples_mch$cells, samples_mch$treatment, samples_mch$time, sep = "_")

d.counts.obs <- list()

for(i in 1:length(counts.obs)){
  d.counts.obs[[i]] = DGEList(counts=counts.obs[[i]], group = group)
  d.counts.obs[[i]]$samples$treatment = samples_mch$treatment
  d.counts.obs[[i]]$samples$time = samples_mch$time
  d.counts.obs[[i]]$samples$cells = samples_mch$cells
}

# use same names 
names(d.counts.obs) <- names(counts.obs)
```

# Barplot of library sizes (un-normalised)
```{r}
pdf(file ="plots/barplots_librarysize.pdf", width = 6, height = 4)
barplot(
  d.counts.obs[["zeb_counts"]]$samples$lib.size,
  names.arg = names.arg,
  las=2, 
  main = "Library sizes (counts)"
  )
barplot(d.counts.obs[["zeb_TCcounts"]]$samples$lib.size,
  names.arg = names.arg,
  las=2, 
  main = "Library sizes(TC counts)"
  )
dev.off()
```

# Analyse different normalisation methods that can be used, for zebrafish and drosophila spike-in
# default is TMM in edgeR
```{r}
# function for plotting normalisation 
BarPlotOfNorm = function(X, Y) {
  ScaleFactors <- X$samples$lib.size * X$samples$norm.factors
  Exp <- round(t(t(X$counts)/ScaleFactors) * mean(ScaleFactors))
  return (Bplot = barplot(log2(Exp+1), main=Y, las=2, names.arg=names.arg))}

# initiate norm lists
d.counts.tmm <- list()
d.counts.UQ <- list()
d.counts.RLE <- list()
d.counts.none <- list()

for(i in 1:length(d.counts.obs)){
    # counts
    d.counts.tmm[[i]] = calcNormFactors(d.counts.obs[[i]], method="TMM")
    d.counts.UQ[[i]] = calcNormFactors(d.counts.obs[[i]], method="upperquartile") #Calculating UpperQuartile normalisation
    d.counts.RLE[[i]] = calcNormFactors(d.counts.obs[[i]], method="RLE")
    d.counts.none[[i]] = calcNormFactors(d.counts.obs[[i]], method="none")

    #Viewing normalisation factors for TMM
    d.counts.tmm[[i]]$samples$norm.factors

    pdf(
        file.path(plots.dir, paste0("barplots_normalisation_", names(d.counts.obs)[[i]], ".pdf")), 
        width = 6, height = 4
        )
    par(mfrow=c(2,2))
    BarPlotOfNorm(d.counts.tmm[[i]], "TMM")
    BarPlotOfNorm(d.counts.UQ[[i]], "UpperQ")
    BarPlotOfNorm(d.counts.RLE[[i]], "RLE")
    BarPlotOfNorm(d.counts.none[[i]], "No normalisation")
    dev.off()
}

#Viewing normalisation factors for TMM and RLE for zebrafish & drsophila counts
norm.table <- data.frame(
  "zeb_counts_tmm" = d.counts.tmm[[1]]$samples$norm.factors,
  "zeb_counts_rle" = d.counts.RLE[[1]]$samples$norm.factors,
  "dros_counts_tmm" = d.counts.tmm[[2]]$samples$norm.factors,
  "dros_counts_rle" = d.counts.RLE[[2]]$samples$norm.factors
)
norm.table

# tmm norm looks best to use in adults and larvae, set names for clarity
names(d.counts.tmm) <- names(counts.obs)
```

# Add gene annotations to counts obj
```{r}
d.counts.obs[["zeb_counts"]]$genes <- genesAnno
```

# MDS plot
```{r}
library(RColorBrewer)
# Setting colours
col.group <- d.counts.obs[["zeb_counts"]]$samples$group
levels(col.group) <- brewer.pal(nlevels(col.group), "Set1") 
col.group <- as.character(col.group)

MDS = plotMDS(
  d.counts.obs[["zeb_counts"]],
  labels=d.counts.obs[["zeb_counts"]]$samples$group, col = col.group
  )
```

# MDS plot TC counts
```{r}
MDS = plotMDS(
  d.counts.obs[["zeb_TCcounts"]],
  labels=d.counts.obs[["zeb_TCcounts"]]$samples$group, col = col.group
  )
```

# Use drosophila TMM as normalisation factor
```{r}
# set zebrafish counts and TCcounts norm factor to dros tmm norm factor
d.counts.tmm[["zeb_counts"]]$samples$norm.factors <- d.counts.tmm[["dros_counts"]]$samples$norm.factors
d.counts.tmm[["zeb_TCcounts"]]$samples$norm.factors <- d.counts.tmm[["dros_counts"]]$samples$norm.factors
```

# MDS for normalised data
```{r}
pdf(file.path(plots.dir, "MDS_larvae_norm_counts.pdf"))
MDS = plotMDS(
  d.counts.tmm[["zeb_counts"]],
  labels=d.counts.tmm[["zeb_counts"]]$samples$group, 
  main = "Larvae normalised counts",
  col = col.group
  )
dev.off()
```

# MDS of normalised TCcounts
```{r}
pdf(file.path(plots.dir, "MDS_larvae_norm_TCcounts.pdf"))
MDS = plotMDS(
  d.counts.tmm[["zeb_TCcounts"]],
  labels=d.counts.tmm[["zeb_TCcounts"]]$samples$group, 
  main = "Larvae normalised TC counts",
  col = col.group
  )
dev.off()
```

# set design and contrasts for DE
```{r}
# design table
design = model.matrix(~0+group, data = d.counts.obs[["zeb_counts"]]$samples)
colnames(design) <- gsub("group", colnames(design), replacement = "")
design

contr.matrix <- makeContrasts(
  APAP3vsDMSO3_WT = WT_UPRT_APAP_3-WT_UPRT_DMSO_3,
  APAP3vsDMSO3_Nrf2KO = Nrf2KO_UPRT_APAP_3-Nrf2KO_UPRT_DMSO_3,
  APAP6vsDMSO6_WT = WT_UPRT_APAP_6-WT_UPRT_DMSO_6,
  APAP6vsDMSO6_Nrf2KO = Nrf2KO_UPRT_APAP_6-Nrf2KO_UPRT_DMSO_6,
  Nrf2KOvsWT_APAP3 = Nrf2KO_UPRT_APAP_3-WT_UPRT_APAP_3, 
  Nrf2KOvsWT_APAP6 = Nrf2KO_UPRT_APAP_6-WT_UPRT_APAP_6, 
  APAP6vsAPAP3_WT = WT_UPRT_APAP_6-WT_UPRT_APAP_3, 
  APAP6vsAPAP3_Nrf2KO = Nrf2KO_UPRT_APAP_6-Nrf2KO_UPRT_APAP_3,
  WT_UPRTvsWT_CFP_DMSO6 = WT_UPRT_DMSO_6-WT_CFP_DMSO_6,
  levels = colnames(design)
  )
contr.matrix
```

EdgeR estimate dispersions (not using in this analysis - using voom instead)
```{r}
# est disp for zebrafish counts
d_zeb_counts=estimateDisp(d.counts.tmm[["zeb_counts"]], design)

# est disp for zebrafish TCcounts
d_zeb_TCcounts=estimateDisp(d.counts.tmm[["zeb_TCcounts"]],design)
```

# start with voom for analysis

# For counts
MA plots
```{r}
v_zeb_counts = voom(d_zeb_counts, design, plot= T)
vfit_zeb_counts <- lmFit(v_zeb_counts, design)
vfit_zeb_counts <- contrasts.fit(vfit_zeb_counts, contrasts=contr.matrix)
efit_zeb_counts <- eBayes(vfit_zeb_counts)
plotSA(efit_zeb_counts)
```

Summary of constrasts
```{r}
summary(decideTests(efit_zeb_counts))
```

Summary of contrasts with logFC greater than 1
```{r}
tfit_zeb_counts <- treat(vfit_zeb_counts, lfc=1)
dt_zeb_counts <- decideTests(tfit_zeb_counts)
summary(dt_zeb_counts)
```

# DE genes
```{r}
# initialise list
contrast.list <- list()

#set and make DE dir
de.dir <- file.path(results.dir, "DE")
dir.create(de.dir)
voom.dir <- file.path(de.dir, "voom")
dir.create(voom.dir)

for(contrast.i in 1:length(colnames(contr.matrix))){
  contrast.list[[contrast.i]] <- topTreat(tfit_zeb_counts, coef=contrast.i, n=Inf)
  contrast.list[[contrast.i]]$ensg <- rownames(contrast.list[[contrast.i]])
  contrast.list[[contrast.i]] <- merge(contrast.list[[contrast.i]], genesAnno, by.x = "ensg", by.y = "zfish_ensembl_gene_id")
  write.table(
    contrast.list[[contrast.i]],
    file.path(voom.dir, paste0(colnames(contr.matrix)[[contrast.i]], "_Counts.txt")),
    sep="\t",
    quote=F,
    row.names = F,
    col.names = T
    )
}

names(contrast.list) <- colnames(contr.matrix)
```

# For TCcounts
MA plots
```{r}
v_zeb_TCcounts = voom(d_zeb_TCcounts, design, plot= T)
vfit_zeb_TCcounts <- lmFit(v_zeb_TCcounts, design)
vfit_zeb_TCcounts <- contrasts.fit(vfit_zeb_TCcounts, contrasts=contr.matrix)
efit_zeb_TCcounts <- eBayes(vfit_zeb_TCcounts)
plotSA(efit_zeb_TCcounts)
```

Summary of contrasts (number of DE genes)
```{r}
summary(decideTests(efit_zeb_TCcounts))
```

Summary of contrasts with logFC greater than 1
```{r}
tfit_zeb_TCcounts <- treat(vfit_zeb_TCcounts, lfc=1)
dt_zeb_TCcounts <- decideTests(tfit_zeb_TCcounts)
summary(dt_zeb_TCcounts)
```

# DE 
```{r}
# initialise list
TC.contrast.list <- list()

for(contrast.i in 1:length(colnames(contr.matrix))){
  TC.contrast.list[[contrast.i]] <- topTreat(tfit_zeb_TCcounts, coef=contrast.i, n=Inf)
  TC.contrast.list[[contrast.i]]$ensg <- rownames(TC.contrast.list[[contrast.i]])
  TC.contrast.list[[contrast.i]] <- merge(TC.contrast.list[[contrast.i]], genesAnno, by.x = "ensg", by.y = "zfish_ensembl_gene_id")
  write.table(
    TC.contrast.list[[contrast.i]],
    file.path(voom.dir, paste0(colnames(contr.matrix)[[contrast.i]], "_TCCounts.txt")),
    sep="\t",
    quote=F,
    row.names = F,
    col.names = T
    )
}

names(TC.contrast.list) <- colnames(contr.matrix)
```

MA plots
```{r}
voom.ma.dir <- file.path(plots.dir, "voom_MAplots")
dir.create(voom.ma.dir)
for(contrast.i in 1:length(TC.contrast.list)){
  pdf(file.path(voom.ma.dir, paste0(colnames(tfit_zeb_TCcounts)[contrast.i], "_MAplot.pdf")))
  plotMD(
    tfit_zeb_TCcounts, 
    column=contrast.i, 
    status=dt_zeb_TCcounts[,contrast.i],
    main=colnames(tfit_zeb_TCcounts)[contrast.i], 
    xlim=c(0,13)
    )
  dev.off()
}
```

# also look at edgeR de for counts
```{r}
gfit_zeb_counts <- glmQLFit(d_zeb_counts, design)
edger.dir <- file.path(de.dir, "edger")
dir.create(edger.dir)
ftest.counts.list <- list()

for(contrast.i in 1:length(colnames(contr.matrix))){

  ftest.counts.list[[contrast.i]] <- glmQLFTest(gfit_zeb_counts, contrast = contr.matrix[ , contrast.i])
  ftest.counts.list[[contrast.i]]$table$ensg <- rownames(ftest.counts.list[[contrast.i]]$table)
  ftest.counts.list[[contrast.i]]$table <- merge(ftest.counts.list[[contrast.i]]$table, genesAnno, by.x = "ensg", by.y = "zfish_ensembl_gene_id")
  # write ftest table
  write.table(
    ftest.counts.list[[contrast.i]]$table,
    file.path(edger.dir, paste0(colnames(contr.matrix)[contrast.i], "_counts_ftest_table.txt")),
    sep = "\t",
    quote = F,
    col.names = T,
    row.names= F
  )
}
```

# for TC counts
```{r}
gfit_zeb_TCcounts <- glmQLFit(d_zeb_TCcounts, design)
ftest.TCcounts.list <- list()
for(contrast.i in 1:length(colnames(contr.matrix))){

  ftest.TCcounts.list[[contrast.i]] <- glmQLFTest(gfit_zeb_TCcounts, contrast = contr.matrix[ , contrast.i])
  ftest.TCcounts.list[[contrast.i]]$table$ensg <- rownames(ftest.TCcounts.list[[contrast.i]]$table)
  ftest.TCcounts.list[[contrast.i]]$table <- merge(ftest.TCcounts.list[[contrast.i]]$table, genesAnno, by.x = "ensg", by.y = "zfish_ensembl_gene_id")
  # write ftest table
  write.table(
    ftest.TCcounts.list[[contrast.i]]$table,
    file.path(edger.dir, paste0(colnames(contr.matrix)[contrast.i], "_TCcounts_ftest_table.txt")),
    sep = "\t",
    quote = F,
    col.names = T,
    row.names= F
  )
}

# add names for referencing
names(ftest.TCcounts.list) <- colnames(contr.matrix)
```

# MA plots
```{r}
edger.ma.dir <- file.path(plots.dir, "edger_MAplots")
dir.create(edger.ma.dir)
for(contrast.i in 1:length(colnames(contr.matrix))){
  pdf(file.path(edger.ma.dir, paste0(colnames(contr.matrix)[contrast.i], "_MAplot.pdf")))
  maPlot(
    logAbundance = ftest.TCcounts.list[[contrast.i]]$table$logCPM,
    logFC = ftest.TCcounts.list[[contrast.i]]$table$logFC,
    main = colnames(contr.matrix)[contrast.i],
    xlab = expression(bold(paste("log"[2], " CPM"))),
    ylab = expression(bold(paste("log"[2], " FC"))),
    col = "black",
    allCol ="red",
    lowess = T
  )
  dev.off()
}
```

# get human orthologs of zebrafish genes
```{r}
Hsensembl=useMart(
  host="https://asia.ensembl.org", 
  biomart='ENSEMBL_MART_ENSEMBL',
  dataset='hsapiens_gene_ensembl'
  )

hum_zeb_link=getLDS(
  attributes = c("ensembl_gene_id", "hgnc_symbol", "entrezgene_id"), 
  mart = Hsensembl,
  attributesL = c("ensembl_gene_id","zfin_id_symbol","entrezgene_id"),
  martL = Drensembl,
  uniqueRows = TRUE
  )

colnames(hum_zeb_link) <- c("human_ensembl_id", "human_symbol", "human_entrez",  "zfish_ensembl_id", "zfin_symbol", "zfish_entrez")
```

# look at gene set enrichment
# start with gprofiler to look for any enrichment in top 100 genes for each contrast
```{r}
library(gprofiler2)
gprof.data.dir <- file.path(results.dir, "gprofiler")
dir.create(gprof.data.dir)

hum.zeb.table <- list()

for(contrast.i in 1:length(colnames(contr.matrix))){
   cont.table <- ftest.TCcounts.list[[contrast.i]]$table
   cont.table <- cont.table[order(cont.table$PValue), ]
   # merge with human genes and save for further analysis
   hum.zeb.table[[contrast.i]] <- merge(cont.table, hum_zeb_link, by.x = "ensg", by.y = "zfish_ensembl_id")
   # check top 100 for enrichment testing
   enrich.table <- hum.zeb.table[[contrast.i]][1:100, ]
   gostres <- gost(
     query = enrich.table$human_ensembl_id, 
     organism = "hsapiens", 
     ordered_query = T, 
     significant = T
     )

   write.table(
     gostres$result[,3:13],
     file.path(gprof.data.dir, paste0(colnames(contr.matrix)[contrast.i], "_top100_func_enrich.txt")),
     sep ="\t",
     quote = F,
     col.names = T,
     row.names = F
   )
}
names(hum.zeb.table) <- colnames(contr.matrix)
```

# use EGSEA on voom data
```{r}
library(EGSEA)

# add human entrez ids and symbols to voom obj
hum_zeb_voom <- merge(
  data.frame("voom_ensg" = rownames(v_zeb_TCcounts$E)), 
  hum_zeb_link, 
  by.x = "voom_ensg", 
  by.y = "zfish_ensembl_id"
  )
# remove dups and entries w/o entrezid or symbol
hum_zeb_voom <- hum_zeb_voom[1:4]
hum_zeb_voom <- hum_zeb_voom[-which(is.na(hum_zeb_voom$human_entrez)), ]
hum_zeb_voom <- hum_zeb_voom[-which(hum_zeb_voom$human_symbol == ""), ]
# removes multiple human genes mapped to same zfish gene, just keeps first zfish gene listed 
# (not sure if this is the ideal way to handle this, but only one human entrez ID can be used per zfish gene)
hum_zeb_voom <- hum_zeb_voom[!duplicated(hum_zeb_voom$voom_ensg), ]

# make voom object limited to human orthologs w/entrez and replace rownames of counts w/entrez IDs
eg_zeb_TCcounts <- v_zeb_TCcounts
eg_zeb_TCcounts$genes <- hum_zeb_voom
keep <- which(rownames(eg_zeb_TCcounts) %in% eg_zeb_TCcounts$genes$voom_ensg)
eg_zeb_TCcounts <- eg_zeb_TCcounts[keep, ]
rownames(eg_zeb_TCcounts$E) <- eg_zeb_TCcounts$genes$human_entrez
# also need to remove multiple zfish genes mapped to same human entrez, using same logic as previous (probably not ideal)
keep.entrez <- which(!duplicated(rownames(eg_zeb_TCcounts)))
eg_zeb_TCcounts <- eg_zeb_TCcounts[keep.entrez, ]

# pre-process and index geneset collections
gs.annots = buildIdx(
  entrezIDs = rownames(eg_zeb_TCcounts$E), 
  species = "human",
  msigdb.gsets = "all", 
  gsdb.gsets ="all", 
  kegg.updated = T
  )

# set params for egsea
hyp.free.methods <- egsea.base()[-c(1, 4, 8, 11)]
egsea.dir <- file.path(results.dir, "egsea")
dir.create(egsea.dir)
hyp.free.dir <- file.path(egsea.dir, "hypothesis_free")
dir.create(hyp.free.dir)

# run egsea and generate html report
gsa = egsea(
  voom.results = eg_zeb_TCcounts,
  contrasts = contr.matrix, 
  gs.annots = gs.annots,
  symbolsMap = eg_zeb_TCcounts$genes[,3:4], 
  baseGSEAs = hyp.free.methods, 
  report.dir = hyp.free.dir, 
  sort.by = "avg.rank", 
  num.threads = 4, 
  report = FALSE
  )

# write out top hits for each contrast in each database
gs.collections <- c("h", "c1", "c2", "c3", "c4", "c5", "c6", "c7", "gsdbdrug", "gsdbdis", "gsdbpath", "gsdbgo", "gsdbreg", "kegg")
for(contrast.i in 1:length(colnames(contr.matrix))){
  for(coll.i in 1:length(gs.collections)){
    topset <- topSets(
      gsa, 
      contrast = contrast.i, 
      gs.label = gs.collections[coll.i], 
      number = 10, 
      names.only = FALSE
      )
    write.table(
      topset,
      file.path(hyp.free.dir, paste0(colnames(contr.matrix)[contrast.i], "_", gs.collections[coll.i], "_top_sets.txt")),
      sep = "\t",
      quote = FALSE,
      col.names = NA,
      row.names = TRUE
    )
  }
}


# now test genesets with hypothesis-driven methods
geneset.dir <- file.path(egsea.dir, "geneset_specific")
dir.create(geneset.dir)
hyp.driven.methods <- egsea.base()[c(1, 4, 8, 11)]

# start w/nrf2 targets geneset, write out results for each comparison
nrf2.dir <- file.path(geneset.dir, "nrf2_targets")
dir.create(nrf2.dir)
nrf2.targets <- scan(
  file.path(nrf2.dir, "nrf2_targets.txt"),
  what = "character"
  )

# limit egsea voom obj to nrf2 genes
keep.nrf2 <- which(eg_zeb_TCcounts$genes$human_symbol %in% nrf2.targets)
eg_nrf2_TCcounts <- eg_zeb_TCcounts[keep.nrf2, ]

nrf2.annots = buildIdx(
  entrezIDs = rownames(eg_nrf2_TCcounts$E), 
  species = "human",
  msigdb.gsets = "all", 
  gsdb.gsets ="all", 
  kegg.updated = TRUE
  )
nrf2.gsa = egsea(
  voom.results = eg_nrf2_TCcounts,
  contrasts = contr.matrix, 
  gs.annots = nrf2.annots,
  symbolsMap = eg_nrf2_TCcounts$genes[,3:4], 
  baseGSEAs = hyp.driven.methods, 
  report.dir = nrf2.dir, 
  sort.by = "avg.rank", 
  num.threads = 4, 
  report = FALSE
  )

for(contrast.i in 1:length(colnames(contr.matrix))){
  for(coll.i in 1:length(gs.collections)){
    topset <- topSets(
      nrf2.gsa, 
      contrast = contrast.i, 
      gs.label = gs.collections[coll.i], 
      number = 10, 
      names.only = FALSE
      )
    write.table(
      topset,
      file.path(nrf2.dir, paste0(colnames(contr.matrix)[contrast.i], "_", gs.collections[coll.i], "_top_sets.txt")),
      sep = "\t",
      quote = FALSE,
      col.names = NA,
      row.names = TRUE
    )
  }
}

# same analyses for ros pathway genes
ros.dir <- file.path(geneset.dir, "ros_pathway")
dir.create(ros.dir)
ros.pathway <- scan(
  file.path(ros.dir, "ros_pathway.txt"),
  what = "character"
  )

# limit egsea voom obj to ros genes, run gsea and write out each result
keep.ros <- which(eg_zeb_TCcounts$genes$human_symbol %in% ros.pathway)
eg_ros_TCcounts <- eg_zeb_TCcounts[keep.ros, ]

ros.annots = buildIdx(
  entrezIDs = rownames(eg_ros_TCcounts$E), 
  species = "human",
  msigdb.gsets = "all", 
  gsdb.gsets ="all", 
  kegg.updated = TRUE
  )
ros.gsa = egsea(
  voom.results = eg_ros_TCcounts,
  contrasts = contr.matrix, 
  gs.annots = ros.annots,
  symbolsMap = eg_ros_TCcounts$genes[,3:4], 
  baseGSEAs = hyp.driven.methods, 
  report.dir = ros.dir, 
  sort.by = "avg.rank", 
  num.threads = 4, 
  report = FALSE
  )

for(contrast.i in 1:length(colnames(contr.matrix))){
  for(coll.i in 1:length(gs.collections)){
    topset <- topSets(
      ros.gsa, 
      contrast = contrast.i, 
      gs.label = gs.collections[coll.i], 
      number = 10, 
      names.only = FALSE
      )
    write.table(
      topset,
      file.path(ros.dir, paste0(colnames(contr.matrix)[contrast.i], "_", gs.collections[coll.i], "_top_sets.txt")),
      sep = "\t",
      quote = FALSE,
      col.names = NA,
      row.names = TRUE
    )
  }
}
```

# make heatmap plots for kegg for each gsa
```{r}
eg.plots.dir <- file.path(plots.dir, "egsea")
dir.create(eg.plots.dir)
heatmap.dir <- file.path(eg.plots.dir, "heatmap")
dir.create(heatmap.dir)

plotSummaryHeatmap(
  gsa, 
  gs.label = "kegg", 
  number = 50, 
  file.name = file.path(heatmap.dir, "hypothesis_free_kegg_heatmap")
  )
plotSummaryHeatmap(
  nrf2.gsa, 
  gs.label = "kegg", 
  number = 50, 
  file.name = file.path(heatmap.dir, "nrf2_kegg_heatmap")
  )
plotSummaryHeatmap(
  ros.gsa, 
  gs.label = "kegg", 
  number = 50, 
  file.name = file.path(heatmap.dir, "ros_kegg_heatmap")
  )
```

# make kegg pathway plot for example comparison
```{r}
pathway.dir <- file.path(eg.plots.dir, "pathway")
dir.create(pathway.dir)

plotPathway(
  gsa, 
  gene.set = "Pathways in cancer", 
  gs.label = "kegg", 
  contrast = 1,
  file.name = file.path(pathway.dir, "APAP3vsDMSO3_WT_kegg_cancer_pathways")
  )
plotPathway(
  gsa, 
  gene.set = "Chemical carcinogenesis", 
  gs.label = "kegg",
  contrast = 1,
  file.name = file.path(pathway.dir, "APAP3vsDMSO3_WT_kegg_chemical_carcinogenesis")
  )
```

# make barcode plot for nrf2 targets and ros pathways
```{r}
barplot.dir <- file.path(eg.plots.dir, "barcode")
dir.create(barplot.dir)
nrf2.bar.dir <- file.path(barplot.dir, "nrf2")
dir.create(nrf2.bar.dir)
ros.bar.dir <- file.path(barplot.dir, "ros")
dir.create(ros.bar.dir)

nrf2.idx <- which(rownames(efit_zeb_TCcounts) %in% eg_nrf2_TCcounts$genes$voom_ensg)
for(contrast.i in 1:length(colnames(contr.matrix))){
  pdf(file.path(nrf2.bar.dir, paste0(colnames(contr.matrix)[contrast.i], "_nrf2_barcode.pdf")))
  barcodeplot(
    efit_zeb_TCcounts$t[,contrast.i], 
    index = nrf2.idx, 
    main = paste0(colnames(contr.matrix)[contrast.i], " NRF2 targets")
    )
  dev.off()
}

ros.idx <- which(rownames(efit_zeb_TCcounts) %in% eg_ros_TCcounts$genes$voom_ensg)
for(contrast.i in 1:length(colnames(contr.matrix))){
  pdf(file.path(ros.bar.dir, paste0(colnames(contr.matrix)[contrast.i], "_ros_barcode.pdf")))
  barcodeplot(
    efit_zeb_TCcounts$t[,contrast.i], 
    index = ros.idx, 
    main = paste0(colnames(contr.matrix)[contrast.i], " ROS pathway")
    )
  dev.off()
}
```

# make contour plot
```{r}
# get TC cpms to test
tc.cpms <- fread(file.path(results.dir, "TCcounts_cpms_filtered_zebrafish_mCherry.txt"))
# select 2 samples to compare (as x&y)
X <- tc.cpms[, c("G3-VT-3680_S21", "H3-VT-3680_S24")]
# remove 0 in either
X <- X[-which(rowSums(X == 0) > 0), ]
# convert to log10 values
X <- log10(X)
# get 2d density estimate for contour overlay
z <- kde2d(unlist(X[,1]), unlist(X[,2]), n = 50)

# plot just lines
plot(unlist(X[,1]), unlist(X[,2]), xlab = "G3-VT-3680_S21", ylab =" H3-VT-3680_S24", pch = 19)
contour(z, lwd = 2, add = TRUE, col = hcl.colors(10, "Spectral"))

# plot with filled in contour
filled.contour(z, xlab = "G3-VT-3680_S21", ylab = "H3-VT-3680_S24")

# plot with additional arguments for adding cor stat, changing line width, and colours
plot(X, xlab = "G3-VT-3680_S21", ylab = "H3-VT-3680_S24", pch = 19, cex = 0.4)
contour(z, lwd = 5, drawlabels = FALSE, nlevels = k, col = my.cols, add = TRUE)
legend("topleft", paste("R=", round(cor(X)[1,2],2)), bty = "n")
```

# save session info
```{r}
sesh.info <- capture.output(sessionInfo())
writeLines(
  sesh.info,
  file.path(project.dir, "session_info", paste0(Sys.Date(), "_slamseq_analysis_2110.txt"))
)
```
